miden-crypto/src/merkle/smt/historical/mod.rs

```diff
@@ -0,0 +1,500 @@
+//! We start with a bootstrap state `Smt`. For sake of brevity, `rev` is short for a `Reversion`
+//! instance.
+//!
+//! I.e. for the follwing call sequence
+//! ```text
+//! # [rev[n] .. rev[1] rev[0]] Smt
+//!
+//! #construction
+//! [] Smt # <- smt
+//!
+//! # apply_mtuations(fwd0)
+//! [rev(fwd0)] (smt + fwd0) = [rev(fwd0)] (smt')
+//!
+//! # apply_mtuations(fwd1)
+//! [rev(fwd0) rev(fwd1)] (smt' + fwd1) = [rev(fwd0) rev(fwd1)] (smt'')
+//! #                                      [2]       [1]         [0] <- historical index
+//! ```
+//!
+//! Now in practice we are very much interested in the key/value leaf pairs, so we cache them
+//! alongside with each reversion.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::{Arc, RwLock, RwLockReadGuard},
+    vec::Vec,
+};
+
+use crate::{
+    EMPTY_WORD, Word,
+    merkle::{
+        EmptySubtreeRoots, InnerNode, LeafIndex, MerkleError, MutationSet, NodeIndex, NodeMutation,
+        SMT_DEPTH, Smt, SmtLeaf, SmtProof, SparseMerklePath, smt::SparseMerkleTree,
+    },
+};
+
+#[cfg(test)]
+mod tests;
+
+#[allow(missing_docs)]
+#[derive(thiserror::Error, Debug)]
+pub enum HistoricalError {
+    #[error(transparent)]
+    MerkleError(#[from] MerkleError),
+}
+
+/// The offset enum representing the offset relative to the `latest` `Smt`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum HistoricalOffset {
+    ReversionsIdx(usize),
+    Latest,
+    TooAncient,
+}
+#[derive(Debug, Clone)]
+struct HistoricalReversion {
+    /// The root at this historical point
+    root: Word,
+    /// The mutations that were applied to reach this state (Arc'd for sharing)
+    /// This is a vector to support potential batching of mutations at a single historical point
+    mutations: Arc<MutationSet<SMT_DEPTH, Word, Word>>,
+    /// Calculating leaves is expense, so we keep a cache, based on the leaf digest.
+    precomputed_leaves: HashMap<LeafIndex<SMT_DEPTH>, SmtLeaf>,
+}
+impl HistoricalReversion {
+    /// Note: Called after `latest` was updated with the latest mutations. The mutations passed in
+    /// to the function is the inverted set.
+    fn from_reversion_mutation_set(
+        inner: &mut InnerState,
+        mutations: MutationSet<SMT_DEPTH, Word, Word>,
+    ) -> Self {
+        let precomputed_leaves = Self::compute_leaves_for_reversion(inner, &mutations);
+        Self {
+            root: mutations.root(), // The root after applying this reversion
+            mutations: Arc::new(mutations),
+            precomputed_leaves,
+        }
+    }
+
+    /// Pre-compute leaves that are being reverted (i.e., their state before the forward mutation).
+    /// These cached leaves will be used when accessing historical views.
+    ///
+    /// The approach: for each affected leaf position, compute the full SmtLeaf
+    /// based on the current state (after mutation) but with the old values from the reversion.
+    fn compute_leaves_for_reversion(
```

Comment:
Optimization idea: Since I expect SMT updates typically affect a small minority of leaves (?), we could optimize `compute_leaves_for_reversion` by adding early pruning logic. For each affected leaf, we could pre-check if any keys in that leaf's range actually changed before reconstructing the full `SmtLeaf`. This could be done by maintaining a `HashMap<LeafIndex, Vec<(Word, Word)>>` of only the mutations, then checking if those mutations actually affect the leaf state before doing expensive leaf reconstruction. In most cases where leaves are unchanged, we could skip the `O(1024)` leaf scanning entirely, reducing complexity from `O(affected_leaves × 1024)` to `O(actual_changed_mutations)`.

Reply:
The primary use-case was usage with `AccountId`s which by definition don't have the x1023 factor, since every account id gets a "full" leaf to itself.